### PR TITLE
chore(kubernetes_logs source): Adds tests for checkpointing regressions in k8s

### PR DESCRIPTION
**Pull request by eeyun**
Signed-off-by: Ian Henry <ianjhenry00@gmail.com>

### CODE REVIEW

**Comment by jszwedko** [comment]
So, looking at CI, it seems like these tests _sometimes_ catch the checkpointing issue, but not consistently. I think they need to be updated to catch it more consistently. https://github.com/timberio/vector/pull/7140 fixes this issue, so the fixes can be undone to test this test.

Also, these tests are great! Testing end-to-end k8s is no easy feat.

**Comment by StephenWakely** [comment]
@jszwedko I can confirm this test does not capture the issues that arose from #7140. Would you be able to guide me through the steps you went through to reproduce that issue manually?

**Comment by jszwedko** [comment]
> @jszwedko I can confirm this test does not capture the issues that arose from #7140. Would you be able to guide me through the steps you went through to reproduce that issue manually?

Sure thing! It was roughly:

1. Start minikube
2. Deploy vector using our daemonset helm chart (`helm install --namespace vector --create-namespace vector timberio/vector-agent --values values.yaml`)
3. Run `flog` as a k8s "job" to generate some fake logs (https://hub.docker.com/r/mingrammer/flog)
4. Observe that vector has ingested the logs (via `kubectl logs` since the default vector config just dumps them to stdout)
5. Restart the vector daemonset (`kubectl rollout restart -n vector daemonset/vector-agent`)
6. Observe that the new vector pod has re-ingested the same logs via `kubectl logs`

This is from memory though. Let me know if that doesn't do it for you.

**Comment by StephenWakely** [comment]
>     5\. Restart the vector daemonset (`kubectl rollout restart -n vector daemonset/vector-agent`)

Curious. That is pretty much what the test is doing, in particular this line. I'll see if I can repro it manually..

**Comment by spencergilbert** [comment]
I've been looking at this today, and as far as I can tell the test doesn't wait/tail the restarted pod properly

File: lib/k8s-e2e-tests/tests/vector-agent.rs [hunk 1]
```diff
@@ -1572,3 +1572,175 @@ async fn host_metrics() -> Result<(), Box<dyn std::error::Error>> {
...
+
+        // Request to stop the flow.
+        FlowControlCommand::Terminate
+    })
+    .await?;
+    assert!(got_marker);
+
+    framework
+        .restart_rollout("test-vector", "daemonset/vector-agent", vec![])
+        .await?;
```
Comment (spencergilbert) [inline lib/k8s-e2e-tests/tests/vector-agent.rs:1651]: This returns when the _rollout_ is successfully received, not when the new pod has become ready. We moved on too quickly (and weren't following the new Vector instance's output) and always passed the test.

File: lib/k8s-e2e-tests/tests/vector-agent.rs [hunk 2]
```diff
@@ -1572,3 +1572,175 @@ async fn host_metrics() -> Result<(), Box<dyn std::error::Error>> {
...
+        if val["message"].eq("MARKER") {
+            panic!("Checkpointed, marker should not be found");
+        };
+
+        // Request to stop the flow.
+        FlowControlCommand::Terminate
+    })
+    .await?;
+    // We need to re-create the log_reader to connect to the new pod
+    //    let mut log_reader = framework.logs("test-vector", "daemonset/vector-agent")?;
```
Comment (spencergilbert) [inline lib/k8s-e2e-tests/tests/vector-agent.rs:1675]: The `log_reader` was still connected to the original pod, not the newly deployed pod, so we need to reconnect to properly trail the output.

File: lib/k8s-e2e-tests/tests/vector-agent.rs [hunk 3]
```diff
@@ -1572,3 +1572,175 @@ async fn host_metrics() -> Result<(), Box<dyn std::error::Error>> {
...
+    //    let mut lines_till_we_give_up: usize = 10000;
```
Comment (spencergilbert) [inline lib/k8s-e2e-tests/tests/vector-agent.rs:1676]: I pulled all of this commented out code from the `pod_filtering` test defined earlier in this file.

**Review by jszwedko — APPROVED** [review]
Nice work!

I think what you have here is good to merge. Ideally we might use a container where we can control its output so that we could cause it to output directly before and after the vector pod restart rather than relying on the sleep timing. Something like https://hub.docker.com/r/hashicorp/http-echo/

**Comment by spencergilbert** [comment]
> Nice work!
> 
> I think what you have here is good to merge. Ideally we might use a container where we can control its output so that we could cause it to output directly before and after the vector pod restart rather than relying on the sleep timing. Something like https://hub.docker.com/r/hashicorp/http-echo/

We'd have to add some additional bits to expose it, but that's definitely a possibility. I agree that making sure we have enough control over the tests is important

**Comment by jszwedko** [comment]
🎉

**Pull request merged**